File: lib/minimal_tebd/exact_diag.py
"""Exact-diagonalisation reference: full ``2**L``-dimensional state evolution.

Only feasible for small systems (L <= ~14 on a laptop), but invaluable as
a correctness check for TEBD.

Two conventions matter and must match ``mps.py``:

1. The basis ordering of the full vector. We use the standard "big-endian"
   convention where the composite index ``k = sum_i s_i * d**(L-1-i)``
   corresponds to the product state ``|s_0 s_1 ... s_{L-1}>``. This matches
   ``np.kron(A, B) = A (x) B`` with ``A`` acting on the *left* site.

2. The bond Hamiltonian layout. ``H_bonds[b]`` has shape ``(d, d, d, d)``
   with indices ``(i_out, j_out, i_in, j_in)``; after ``.reshape(d*d, d*d)``
   it matches ``np.kron`` on the two-site subspace in the same big-endian order.
"""

import numpy as np
from scipy.linalg import expm


def bond_model_to_full_hamiltonian(model):
    """Assemble the full ``d**L x d**L`` Hamiltonian from bond Hamiltonians.

    ``H = sum_b I^{(x)b} (x) H_b (x) I^{(x)(L-b-2)}``
    """
    L, d = model.L, model.d
    dim = d ** L
    H = np.zeros((dim, dim), dtype=complex)
    I_d = np.eye(d)
    for b, h in enumerate(model.H_bonds):
        left = np.eye(d ** b)
        right = np.eye(d ** (L - b - 2))
        h_mat = h.reshape(d * d, d * d)
        H += np.kron(np.kron(left, h_mat), right)
    return H


def _single_site_op_full(op, i, L, d=2):
    """Embed a ``(d, d)`` single-site operator at site ``i`` into the full space."""
    return np.kron(np.kron(np.eye(d ** i), op), np.eye(d ** (L - i - 1)))


def _two_site_op_full(op_A, op_B, i, j, L, d=2):
    """Embed ``op_A_i op_B_j`` (with ``i != j``) into the full space."""
    if i == j:
        return _single_site_op_full(op_A @ op_B, i, L, d)
    # Sort so i < j for the Kron assembly, but keep track of which op goes where.
    if i < j:
        lo, op_lo, hi, op_hi = i, op_A, j, op_B
    else:
        lo, op_lo, hi, op_hi = j, op_B, i, op_A
    I = np.eye(d)
    parts = [I] * L
    parts[lo] = op_lo
    parts[hi] = op_hi
    out = parts[0]
    for k in range(1, L):
        out = np.kron(out, parts[k])
    return out


class ExactEvolution:
    """Full-state-vector real-time (or imaginary-time) evolution.

    Uses the dense Hamiltonian matrix and a cached ``expm(-1j * dt * H)`` so
    repeated ``step()`` calls are cheap after the first.

    Parameters
    ----------
    model : object with ``L``, ``d``, ``H_bonds``
    state_vector : ndarray of length ``d**L``
        Initial state, will be evolved in place by ``step()``.
    dt : float
        Time step.
    type_evo : {'real', 'imag'}
    """

    def __init__(self, model, state_vector, dt, type_evo='real'):
        self.model = model
        self.L = model.L
        self.d = model.d
        self.psi = np.asarray(state_vector, dtype=complex).copy()
        self.dt = float(dt)
        self.type_evo = type_evo
        self.evolved_time = 0.0
        self.H = bond_model_to_full_hamiltonian(model)
        coeff = -1j if type_evo == 'real' else -1.0
        self.U_dt = expm(coeff * self.dt * self.H)

    # ---- time evolution -------------------------------------------------

    def step(self):
        """Advance by one ``dt``."""
        self.psi = self.U_dt @ self.psi
        if self.type_evo == 'imag':
            self.psi /= np.linalg.norm(self.psi)
        self.evolved_time += self.dt

    # ---- constructors ---------------------------------------------------

    @classmethod
    def from_product_state(cls, model, states, dt, type_evo='real'):
        """Initial state ``|states[0] states[1] ... states[L-1]>``."""
        L, d = model.L, model.d
        assert len(states) == L
        idx = 0
        for s in states:
            idx = idx * d + s
        psi = np.zeros(d ** L, dtype=complex)
        psi[idx] = 1.0
        return cls(model, psi, dt, type_evo=type_evo)

    # ---- observables ----------------------------------------------------

    def expectation_value(self, op):
        """Array ``<op_i>`` of length ``L``."""
        from .mps import _resolve_op
        op_mat = _resolve_op(op)
        out = np.empty(self.L, dtype=complex)
        for i in range(self.L):
            O = _single_site_op_full(op_mat, i, self.L, self.d)
            out[i] = np.vdot(self.psi, O @ self.psi)
        return np.real_if_close(out)

    def correlation_function(self, op_A, op_B):
        """Matrix ``C[i, j] = <op_A_i op_B_j>`` of shape ``(L, L)``."""
        from .mps import _resolve_op
        A = _resolve_op(op_A)
        B = _resolve_op(op_B)
        C = np.zeros((self.L, self.L), dtype=complex)
        for i in range(self.L):
            for j in range(self.L):
                O = _two_site_op_full(A, B, i, j, self.L, self.d)
                C[i, j] = np.vdot(self.psi, O @ self.psi)
        return np.real_if_close(C)

    def entanglement_entropy(self):
        """Von-Neumann entropy at each of the ``L - 1`` interior bonds."""
        out = np.zeros(self.L - 1)
        for b in range(1, self.L):
            # Reshape as (d**b, d**(L-b)) and SVD: Schmidt values on bond b.
            M = self.psi.reshape(self.d ** b, self.d ** (self.L - b))
            s = np.linalg.svd(M, compute_uv=False)
            s = s[s > 1e-20]
            s2 = s * s
            out[b - 1] = -np.sum(s2 * np.log(s2))
        return out

File: lib/minimal_tebd_c/core/exact_diag.cpp
#include "exact_diag.h"

#include <algorithm>
#include <cassert>
#include <cmath>
#include <cstring>
#include <vector>

#include "linalg.h"

// ===========================================================================
// Kronecker product helpers for building full-space operators
// ===========================================================================

// In-place Kronecker: out = A kron B.
// A is (ra x ca), B is (rb x cb), out is (ra*rb x ca*cb).
static void kron_inplace(const Cdouble* A, int ra, int ca, const Cdouble* B,
                         int rb, int cb, Cdouble* out) {
  kron(A, ra, ca, B, rb, cb, out);
}

// Identity matrix of size n, heap-allocated.
static std::vector<Cdouble> eye(int n) {
  std::vector<Cdouble> I(n * n, 0.0);
  for (int i = 0; i < n; ++i) I[i * n + i] = 1.0;
  return I;
}

// Kronecker product returning a new vector.
static std::vector<Cdouble> kron_vec(const Cdouble* A, int ra, int ca,
                                      const Cdouble* B, int rb, int cb) {
  int rr = ra * rb, cc = ca * cb;
  std::vector<Cdouble> out(rr * cc);
  kron(A, ra, ca, B, rb, cb, out.data());
  return out;
}

// ===========================================================================
// Build full Hamiltonian
// ===========================================================================

void bond_model_to_full_hamiltonian(const BondModel& model, Cdouble* H_full) {
  int L = model.L, d = model.d;
  int dim = 1;
  for (int i = 0; i < L; ++i) dim *= d;
  std::memset(H_full, 0, dim * dim * sizeof(Cdouble));

  for (int b = 0; b < L - 1; ++b) {
    // H_bond reshaped to (d*d x d*d).
    const Cdouble* h_mat = model.H_bonds[b].data;
    int d2 = d * d;

    // left = I_{d^b}
    int dim_left = 1;
    for (int i = 0; i < b; ++i) dim_left *= d;
    // right = I_{d^(L-b-2)}
    int dim_right = 1;
    for (int i = 0; i < L - b - 2; ++i) dim_right *= d;

    auto I_left = eye(dim_left);
    auto I_right = eye(dim_right);

    // temp = kron(I_left, h_mat)
    auto temp = kron_vec(I_left.data(), dim_left, dim_left, h_mat, d2, d2);
    int tr = dim_left * d2;
    int tc = dim_left * d2;

    // full = kron(temp, I_right)
    auto full = kron_vec(temp.data(), tr, tc, I_right.data(), dim_right,
                         dim_right);

    // Add to H_full.
    for (int i = 0; i < dim * dim; ++i) H_full[i] += full[i];
  }
}

// ===========================================================================
// Single-site and two-site operators in full space
// ===========================================================================

void single_site_op_full(const Cdouble* op, int i, int L, int d,
                         Cdouble* out) {
  int dim = 1;
  for (int k = 0; k < L; ++k) dim *= d;

  int dim_left = 1;
  for (int k = 0; k < i; ++k) dim_left *= d;
  int dim_right = 1;
  for (int k = 0; k < L - i - 1; ++k) dim_right *= d;

  auto I_left = eye(dim_left);
  auto I_right = eye(dim_right);
  auto temp = kron_vec(I_left.data(), dim_left, dim_left, op, d, d);
  int tr = dim_left * d;
  auto full =
      kron_vec(temp.data(), tr, tr, I_right.data(), dim_right, dim_right);
  std::copy(full.begin(), full.end(), out);
}

void two_site_op_full(const Cdouble* opA, const Cdouble* opB, int i, int j,
                      int L, int d, Cdouble* out) {
  int dim = 1;
  for (int k = 0; k < L; ++k) dim *= d;

  if (i == j) {
    // A @ B at site i.
    Cdouble AB[4];
    for (int p = 0; p < d; ++p)
      for (int q = 0; q < d; ++q) {
        AB[p * d + q] = 0;
        for (int r = 0; r < d; ++r) AB[p * d + q] += opA[p * d + r] * opB[r * d + q];
      }
    single_site_op_full(AB, i, L, d, out);
    return;
  }

  // General case: build as tensor product with identities.
  int lo = std::min(i, j), hi = std::max(i, j);
  const Cdouble* op_lo = (i < j) ? opA : opB;
  const Cdouble* op_hi = (i < j) ? opB : opA;

  auto I_d = eye(d);
  // Build the chain of kron products: I^lo x op_lo x I^(hi-lo-1) x op_hi x I^(L-hi-1)
  // Start with identity of size d^0 = 1.
  std::vector<Cdouble> result = {1.0};
  int cur_dim = 1;
  for (int k = 0; k < L; ++k) {
    const Cdouble* mat;
    if (k == lo)
      mat = op_lo;
    else if (k == hi)
      mat = op_hi;
    else
      mat = I_d.data();
    auto next = kron_vec(result.data(), cur_dim, cur_dim, mat, d, d);
    cur_dim *= d;
    result = std::move(next);
  }
  std::copy(result.begin(), result.end(), out);
}

// ===========================================================================
// ExactEvolution
// ===========================================================================

ExactEvolution::ExactEvolution(const BondModel& model,
                               const Cdouble* state_vector, double dt,
                               int type_evo)
    : L(model.L), d(model.d), dt(dt), type_evo(type_evo), evolved_time(0.0) {
  dim = 1;
  for (int i = 0; i < L; ++i) dim *= d;

  psi = new Cdouble[dim];
  std::copy(state_vector, state_vector + dim, psi);

  H = new Cdouble[dim * dim];
  bond_model_to_full_hamiltonian(model, H);

  Cdouble coeff = (type_evo == 0) ? Cdouble(0, -1) : Cdouble(-1, 0);
  std::vector<Cdouble> scaled_H(dim * dim);
  for (int i = 0; i < dim * dim; ++i) scaled_H[i] = coeff * dt * H[i];

  U_dt = new Cdouble[dim * dim];
  matrix_expm(scaled_H.data(), dim, U_dt);
}

ExactEvolution::~ExactEvolution() {
  delete[] psi;
  delete[] H;
  delete[] U_dt;
}

ExactEvolution::ExactEvolution(ExactEvolution&& other) noexcept
    : L(other.L),
      d(other.d),
      dim(other.dim),
      psi(other.psi),
      H(other.H),
      U_dt(other.U_dt),
      dt(other.dt),
      evolved_time(other.evolved_time),
      type_evo(other.type_evo) {
  other.psi = nullptr;
  other.H = nullptr;
  other.U_dt = nullptr;
}

ExactEvolution ExactEvolution::from_product_state(const BondModel& model,
                                                   const int states[],
                                                   double dt, int type_evo) {
  int L = model.L, d = model.d;
  int dim = 1;
  for (int i = 0; i < L; ++i) dim *= d;

  // Compute index: k = sum_i states[i] * d^(L-1-i)
  int idx = 0;
  for (int i = 0; i < L; ++i) idx = idx * d + states[i];

  std::vector<Cdouble> psi(dim, 0.0);
  psi[idx] = 1.0;

  return ExactEvolution(model, psi.data(), dt, type_evo);
}

void ExactEvolution::step() {
  // psi = U_dt @ psi
  std::vector<Cdouble> new_psi(dim, 0.0);
  zgemv(U_dt, dim, psi, new_psi.data(), dim, dim);
  std::copy(new_psi.begin(), new_psi.end(), psi);

  if (type_evo == 1) {
    // Normalize for imaginary-time evolution.
    double nrm = 0.0;
    for (int i = 0; i < dim; ++i) nrm += std::norm(psi[i]);
    nrm = std::sqrt(nrm);
    for (int i = 0; i < dim; ++i) psi[i] /= nrm;
  }
  evolved_time += dt;
}

void ExactEvolution::expectation_value(const Cdouble* op, double out[]) const {
  std::vector<Cdouble> O_full(dim * dim);
  for (int i = 0; i < L; ++i) {
    single_site_op_full(op, i, L, d, O_full.data());
    // <psi| O |psi> = psi^H @ O @ psi
    std::vector<Cdouble> O_psi(dim, 0.0);
    zgemv(O_full.data(), dim, psi, O_psi.data(), dim, dim);
    Cdouble val = 0.0;
    for (int k = 0; k < dim; ++k) val += std::conj(psi[k]) * O_psi[k];
    out[i] = std::real(val);
  }
}

void ExactEvolution::correlation_function(const Cdouble* opA,
                                          const Cdouble* opB,
                                          double out[]) const {
  std::vector<Cdouble> O_full(dim * dim);
  for (int i = 0; i < L; ++i) {
    for (int j = 0; j < L; ++j) {
      two_site_op_full(opA, opB, i, j, L, d, O_full.data());
      std::vector<Cdouble> O_psi(dim, 0.0);
      zgemv(O_full.data(), dim, psi, O_psi.data(), dim, dim);
      Cdouble val = 0.0;
      for (int k = 0; k < dim; ++k) val += std::conj(psi[k]) * O_psi[k];
      out[i * L + j] = std::real(val);
    }
  }
}

void ExactEvolution::entanglement_entropy(double out[]) const {
  for (int b = 1; b < L; ++b) {
    int dim_left = 1;
    for (int i = 0; i < b; ++i) dim_left *= d;
    int dim_right = dim / dim_left;

    // Reshape psi as (dim_left x dim_right) matrix and SVD.
    int m = dim_left, n = dim_right;
    int mn = std::min(m, n);

    std::vector<Cdouble> U(m * mn);
    std::vector<double> S(mn);
    std::vector<Cdouble> Vt(mn * n);

    if (m >= n) {
      svd(psi, m, n, n, U.data(), S.data(), Vt.data());
    } else {
      // Transpose psi matrix.
      std::vector<Cdouble> psi_t(n * m);
      for (int i = 0; i < m; ++i)
        for (int j = 0; j < n; ++j) psi_t[j * m + i] = psi[i * n + j];
      std::vector<Cdouble> U_t(n * mn);
      std::vector<Cdouble> Vt_t(mn * m);
      svd(psi_t.data(), n, m, m, U_t.data(), S.data(), Vt_t.data());
    }

    double ent = 0.0;
    for (int k = 0; k < mn; ++k) {
      double s = S[k];
      if (s > 1e-20) {
        double s2 = s * s;
        ent -= s2 * std::log(s2);
      }
    }
    out[b - 1] = ent;
  }
}
